predicate/solver/teleport.py

```diff
@@ -207,22 +212,92 @@ def collect_like(self, other: ast.Predicate):
         return [r for r in self.rules if r.__class__ == other.__class__]
 
 
-# TODO: not really sure how I want to structure this logic
-#       as I'd like to keep the logic for non-teleport specific ast elements close to their
-#       definitions, but I am not sure if this is doable without a lot of complexity as the output
-#       format is tied to whatever builds on top of the general ast
-#       this might do for now
-def transform_expr(predicate):
+# t_expr transforms a predicate-lang expression into a Teleport predicate expression which can be evaluated.
+def t_expr(predicate):
+    # special-case predicate subclasses like Node to get their inner expression
     if isinstance(predicate, ast.Predicate):
-        return transform_expr(predicate.expr)
+        if hasattr(predicate, "scope"):
+            # if the predicate has an evaluation scope for which it is only enabled,
+            # case it's expression within a `scope` call to allow defaulting in the expression evaluator
+            return f'scope({t_expr(predicate.expr)}, "{predicate.scope}")'
```

Comment:
This part is not 100% clear to me, can you elaborate why do we need scopes here?

Reply:
Ah, sure. Let's use the allow rules as an example. An allow section is composed of one or more coarse grained rules applicable to a specific scenario as I understand it, notably `Node` in current predicate code. That is, the following predicate DSL is a self-contained rule that we need to evaluate:
```py
Node(
    ((Node.login == User.name) & (User.name != "root"))
    | (User.traits["team"] == ("admins",))
)
```
When assembling these both for the Z3 theorem and when generating the Teleport predicate resource, these rules are merged together with the `OR` logical operator to form a final expression:
```py
# Policy.export
if self.allow.rules:
    allow_rules = functools.reduce(operator.or_, self.allow.rules)
    out["spec"]["allow"] = t_expr(allow_rules)
```
When Teleport later evaluates the allow expression to determine RBAC to say, an SSH node. We only want to think about the evaluation results of any Node rules. If we support say K8S via a `class KubeCluster` rule similar to `class Node`, we may not fill in all the values needed for that subexpression to evaluate correctly, thus it may evaluate to `true`, which may cause the overall allow expression to evaluate to `true` due to the `OR` operator. Thus, we want to selectively avoid evaluating/discard the result of any scopes we aren't interested in. We still want to use an OR rule since we'd like to apply that if multiple `Node` rules are present.

Using a wrapper function like `scope` allows Teleport to discard or avoid evaluating a subexpression when the result doesn't make sense. There are other ways to get this sort of functionality but after some playing around this felt logical for the time being. We could just as easily rename this general `scope()` function to say `Node()` and define one per relevant scope if that seems better to you.

Reply:
OK, I understand now.

Let's do this instead:

```yaml
kind: policy
metadata:
  name: access
spec:
  allow: 
       node: (node.login == user.name) && (!(user.name == "root"))) || (user.traits["team"] == ["admins"])
```

Pretty much the same as `scope`, but read better. Let's also replace scope property with a decorator

```python
@serialize(scope)
class Node(ast.Predicate):
    pass
```

Behind the scenes you can add a property to the class, but we won't have to repeat ourselves and you can derive it from the class name